REP-2008 - ROS 2 Hardware Acceleration Architecture and Conventions (#324)
ros-infrastructure/rep: rep-2008.rst

                    |                |          |    CPU       |
                    +----+------^--^-+          |              |
                         |      |  |            +-------+------+
                         |      |  |                    |
                         +------+  |                    |
                           LTTng   +--------------------+
                                       re-instrument

This can be summarized as follows:

- 1. instrument the core components of a ROS 2 application using `LTTng <https://lttng.org/>`_ and trace its execution to obtain information about its time bottlenecks. Refer to `ros2_tracing <https://gitlab.com/ros-tracing/ros2_tracing>`_ for tools, documentation and ROS 2 core layers tracepoints;

- 2. trace, benchmark and analyze the ROS 2 application on the CPU to establish a compute baseline; Determine which functions are subject to be hardware accelerated.
  
- 3. develop a hardware acceleration kernels for those functions identified before, and optimize the dataflow across Nodes

    - 3.1 accelerate computations at the Node or Component level for each one of those functions identified in 2. as good candidates.

>     - 3.2 accelerate inter-Node exchanges and reduce the overhead of the ROS 2 message-passing system across all its abstraction layers.

Author: ```suggestion
    - 3.2 accelerate inter-Node exchanges and reduce the overhead of the ROS 2 message-passing system across all its abstraction layers. To do so, consider leveraging REP-2007 and REP-2009.
```

The outcome of this REP should be that maintainers who want to leverage hardware acceleration in their packages, can do so with consistent guidelines and with support across multiple hardware acceleration technologies (FPGAs and GPUs initially, but extending towards more accelerators in the future) by following the conventions set. This way, maintainers will be able to create ROS 2 packages with support for hardware acceleration that can run across accelerators. In turn, as silicon vendors add support for their technologies and accelerators, ROS 2 packages aligned with this REP will automatically gain support for these new accelerators.

The guidelines in here provide a ROS 2-centric open architecture for hardware acceleration, which silicon vendors can decide to adopt when engaging with the ROS 2 community.

Antigoals
^^^^^^^^^

The motivation behind this REP does **not** include:

* Dictating policy implementation specifics on maintainers to achieve hardware acceleration and/or create acceleration kernels (e.g use of CUDA, HLS, OpenCL, or similar)

>   * Policy requirements about which framework to use (e.g. ROCm, JetPack, etc.) or wihch specific C++ preprocessor instructions are *intentionally generic* and to remain technology-agnostic.

Reviewer: ```suggestion
  * Policy requirements about which framework to use (e.g. ROCm, JetPack, etc.) or which specific C++ preprocessor instructions are *intentionally generic* and to remain technology-agnostic.
```
Also, this bullet point is worded a bit awkward imo.

Author: Thanks!

..   * - 
     - ``ROS_XRT``
     -  This CMake variable will evaluate to ``True`` if the Xilinx Runtime (XRT) Library  is enabled in the build system. This variable is meant to be used by the maintainers to conditionally compile their ROS 2 packages for acceleration kernels that leverage XRT.

Through ``ament_hardware_acceleration`` and technology-specific specializations (like ``ament_vitis``), the ROS 2 build system is automatically enhanced to support producing  acceleration kernel and related artifacts as part of the build process when invoking ``colcon build``. To facilitate the work of maintainers, this additional functionality is configurable through ``mixins`` that can be added to the build step of a ROS 2 workspace, triggering all the hardware acceleration logic only when appropriate (e.g. when ``--mixin kv260`` is appended to the build effort, it'll trigger the build of kernels targeting the KV260 hardware solution). For a reference implementation of these enhacements, refer to `ament_vitis <https://github.com/ros-acceleration/ament_vitis/>`_.

> In turn, additional extensions to the existing CMake macros might be proposed which would allow to support more technologies and hardware solutions. For example, ``ament_vitis`` provides a ``vitis_acceleration_kernel`` macro that can be used to generate kernels for the Xilinx Vitis platform. Similarly, ``ament_rocm`` could provide a ``rocm_acceleration_kernel`` macro that can be used to generate kernels for the AMD ROCm platform. This way, maintainers can choose to use the generic ``acceleration_kernel`` macro, or instead use the technology-specific macros to target specific hardware solutions. Also, future extensions to the ``ament_hardware_acceleration`` package could be proposed to support the use of accelerators in binary formats, instead of build them from source code. This would allow to support accelerators which are are not fully integrated into ROS 2 through their corresponding technology libraries (e.g. FPGAs that are not supported by Vitis).

Reviewer: ```suggestion
In turn, additional extensions to the existing CMake macros might be proposed which would allow to support more technologies and hardware solutions. For example, ``ament_vitis`` provides a ``vitis_acceleration_kernel`` macro that can be used to generate kernels for the Xilinx Vitis platform. Similarly, ``ament_rocm`` could provide a ``rocm_acceleration_kernel`` macro that can be used to generate kernels for the AMD ROCm platform. This way, maintainers can choose to use the generic ``acceleration_kernel`` macro, or instead use the technology-specific macros to target specific hardware solutions. Also, future extensions to the ``ament_hardware_acceleration`` package could be proposed to support the use of accelerators in binary format instead of source builds. This would allow to support accelerators which are are not fully integrated into ROS 2 through their corresponding technology libraries (e.g. FPGAs that are not supported by Vitis).
```

                           LTTng   +--------------------+
                                       re-instrument

The following proposes a methodology to analyze a ROS 2 application and design appropriate acceleration:

1. instrument both the core components of ROS 2 and the target kernels using `LTTng <https://lttng.org/>`_. Refer to `ros2_tracing <https://gitlab.com/ros-tracing/ros2_tracing>`_ for tools, documentation and ROS 2 core layers tracepoints;
2. trace and benchmark the kernels on the CPU to establish a compute baseline;
3. develop a hardware accelerated implementation on alternate hardware (e.g., GPU, FPGA, etc):

   - **3.1** accelerate computations at the Node or Component level for each one of those identified in **2.** as good candidates.
   - **3.2** accelerate inter-Node exchanges and reduce the overhead of the ROS 2 message-passing system across all its abstraction layers.

4. trace, benchmark against the CPU baseline, and improve the accelerated implementation.

The proposed ROS 2 methodology for hardware acceleration is demonstrated in [7]_ and [8]_.

.. _acceleration examples:

Acceleration examples
^^^^^^^^^^^^^^^^^^^^^

> For the sake of illustrating maintainers and consumers how to build their own acceleration kernels and guarantee interoperability across technologies, a ROS 2 meta-package named `acceleration_examples <https://github.com/ros-acceleration/acceleration_examples>`_ will be maintained and made available. This meta-package will contain various packages with simple common acceleration examples. Each one of these examples should support all hardware acceleration solutions complying with this REP.

Reviewer 2: I think the requirement that all examples must work on all supported solutions is too strict. 
This requirement essentially means that each hardware solution provider has to adapt all examples to target its own solution, A part from the huge overhead this will impose on eventual providers,  I am not sure this is even/easily feasible. 
Rather, It is reasonable to expect that each hardware solution has at least one example in this package.

Author: I'm ok relaxing the wording in here. Moreover, "*full*" compliance with `acceleration_examples` is somewhat covered at https://github.com/ros-infrastructure/rep/pull/324/files#diff-f230b6aa06d86bf594d8e431300e453ad7343e8f4b1932252b6d36c62a8b5e0aR398, which requires *official* boards to support all examples. I think it's useful for the user to know this information.

> In turn, a CI system will be set to build periodically and for every commit the meta-package.

Author: See https://github.com/ros-infrastructure/rep/pull/324#issuecomment-1188634661 for further insight into the CI efforts. More details to be brought up in upcoming WG meetings and discusses as part of the group.

.. _capabilities:

Capabilities
^^^^^^^^^^^^

The following list describes the hardware acceleration capabilities that hardware acceleration vendors must consider when connecting their firmware and technology solutions to the ROS 2 ecosystem.

.. _Kernel Levels Capabilities:

1. **Kernel Levels:**

   .. _1.i:

   i. Must provide at least one package producing `Level I kernels`_

   .. _1.ii:

   ii. Must provide at least one package producing `Level II kernels`_

   .. _1.iii:

   iii. Must provide at least one package producing `Level III kernels`_

.. _Build System Enhancements:

2. **Build System Enhancements:**

   .. _2.i:

>    i. Must have an ``ament_<technology>`` (e.g. ``ament_vitis``) package that integrates technology-specific frameworks and software platforms into the ROS 2 build system through a series of CMake macros that maintainers can use to generate acceleration kernels from their `CMakeLists.txt` files.

Reviewer 2: I think the requirement these two  requirements (i  and ii) should be optional.
I am not familiar with cuda, but I think it is now supported by new versions of cmake, so I think one can do without ament_cuda for example.

For GPUs, with opencl, you can runtime compile/link programs. In this case as well there is no need for ament_<technology>

This comment is to be taken with a grain of salt

Author: What we're trying to achieve in here is that the same package meta-data (i.e. the `CMakeLists.txt`) contains logic that allows producing acceleration kernels for various accelerators without having to modify it. This is achieved by integrating acceleration frameworks into `ament` build system via CMake macros that abstract away the specifics of each technology. E.g.:

- `ament_acceleration`
  - `ament_vitis`
  - `ament_cuda`
  - `ament_<whatever>`

If newer versions of CUDA can be easily integrated with CMake directly that's fantastic, but we still need the "*ROS build system alignment*". That's what this requirement in here captures. I think this is important.

Pillar III - firmware
^^^^^^^^^^^^^^^^^^^^^

The third pillar is firmware. Represented by the abstract ``acceleration_firmware`` package, it is meant to provide firmware artifacts for each supported technology so that the kernels can be compiled against them, simplifying the process for consumers and maintainers, and further aligning with the ROS typical development flow. 

Each ROS 2 workspace can have one or multiple firmware packages deployed. The selection of the active firmware in the workspace is performed by the ``colcon acceleration select`` subverb (pillarII_). To get a technology solution aligned with this REP's architecture, each vendor should provide and maintain an ``acceleration_firmware_<solution>`` package specialization that delivers the corresponding artifacts in line with its supported categories_ and capabilities_. Firmware artifacts should be deployed at ``<ros2_workspace_path>/acceleration/firmware/<solution>`` and be ready to be used by the ROS 2 build system extensions at (pillarI_) . For a reference implementation of specialized vendor firmware package, see `acceleration_firmware_kv260 <https://github.com/ros-acceleration/acceleration_firmware_kv260>`_. 

> By splitting vendors across packages with a common front-end (``acceleration_firmware``), consumers and maintainers can easily switch between hardware acceleration solutions and 

Author: Thanks for this @gbalke.

.. _specification:

Specification
=============

To drive the creation, maintenance and testing of acceleration kernels in ROS 2 packages that are agnostic to the computing target (i.e. consider support for edge, workstation, data center or cloud targets) and technology-agnostic (considers initial support for FPGAs and GPUs), this REP builds on top of open standards. Particularly, ``OpenCL 1.2`` ([2]_, [3]_) is encouraged for a well established standardized interoperability between the host-side (CPU) and the acceleration kernel. Unless stated otherwise, the hardware acceleration terminology used in this document follows the OpenCL nomenclature for hardware acceleration.

> A ROS 2 package supports hardware acceleration if it provides support for at least one of the *supported hardware acceleration solutions* that comply with this REP.

Author: I like this idea. I think there's a tradeoff of how much we're asking from maintainers, but this doesn't sound much to me and can be merged into the recommendations at the bottom of the document.

I'd say, let's bring those ideas there.

.. _acceleration examples:

Acceleration examples
^^^^^^^^^^^^^^^^^^^^^

For the sake of illustrating maintainers and consumers how to build their own acceleration kernels and guarantee interoperability across technologies, a ROS 2 meta-package named `acceleration_examples <https://github.com/ros-acceleration/acceleration_examples>`_ will be maintained and made available. This meta-package will contain various packages with simple common acceleration examples. Each one of these examples should support all hardware acceleration solutions complying with this REP.

In turn, a CI system will be set to build periodically and for every commit the meta-package.

.. _capabilities:

Capabilities
^^^^^^^^^^^^

The following list describes the hardware acceleration capabilities that hardware acceleration vendors must consider when connecting their firmware and technology solutions to the ROS 2 ecosystem.

> 

Reviewer 3: Github is having a bad time rendering this section in the preview. I tested it out in another rst renderer and it turned out just fine so I'm guessing it's just a bug.
![image](https://user-images.githubusercontent.com/11024792/135360352-0144a84a-6cb4-4bff-a8e9-16a4ac7f5c23.png)

Author: Thanks for looking at this. 

I only iterated locally where I get a nice rendering using `rep2html.py` script. I can confirm it renders HTML fine for me as well.

      - 
      - 
      -     
    * - 5.i_
      - ✓
      - ✓
      - ●
      - 
    * - Testing and CI
      - 
      - 
      - 
      -     
    * - 6.i_
      - ✓
      - ●
      - ○
      - 

Reference Implementation and recommendations
============================================

Reference implementations complying with this REP and extending the ROS 2 build system and tools for hardware acceleration are available at the `Hardware Accelerationg WG GitHub organization <https://github.com/ros-acceleration>`_. This includes also the abstraction layer ``ament_acceleration`` and firmware from vendor specializalizations like `ament_vitis <https://github.com/ros-acceleration/ament_vitis/>`_.

For additional implementations and recommendations, check out the `Hardware Accelerationg WG GitHub organization <https://github.com/ros-acceleration>`_.

.. _labels for maintainers:

Labels for maintainers
^^^^^^^^^^^^^^^^^^^^^^
Maintainers are encouraged to mark their packages including hardware acceleration with a label that indicates so, for each one of the hardware acceleration technology solutions they support. This REP proposes that such indication takes the following shape:

> .. image:: https://img.shields.io/badge/hardware_acceleration-KV260-ec1c24.svg

Reviewer 3: I just browsed some packages and don't see these labels commonly used within the ROS ecosystem. I'm not sure if there's a precedent for this sort of thing in REPs.

Author: You're right, there's no precedent (that I know at least) of using *shields* in REPs.

That said, they're widely used in some of the most popular packages (e.g. [navigation2](https://github.com/ros-planning/navigation2), [moveit2](https://github.com/ros-planning/moveit2), [ros2_control](https://github.com/ros-controls/ros2_control), to mention a few). Given that most maintainers are picking *shields*, I believe it's somewhat fair to encourage them to do so. That's what's attempt in here.

Author: I believe we've addressed this by now since the content has dissappeared from the REP.

    +-------------+             |                      |                            |
    | rmw_adapter |          Pillar I              Pillar II                    Pillar III
    +-------------+
    

.. _pillarI:

Pillar I - Extensions to ament
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The first pillar represents extensions of the ``ament`` ROS 2 build system. These CMake extensions help achieve the objective of simplifying the creation of acceleration kernels. By providing an experience and a syntax similar to other ROS 2 libraries targeting CPUs, maintainers will be able to integrate kernels into their packages easily. The ``ament_acceleration`` ROS 2 package abstracts the build system extensions from technology-specific frameworks and software platforms. This allows to easily support hardware acceleration across FPGAs and GPUs while using the same syntax, simplifying the work of maintainers. The code listing below provides an example that instructs the ``CMakeLists.txt`` file of a ROS 2 package to build a ``vadd`` kernel with the corresponding sources and configuration:

::

    acceleration_kernel(
      NAME vadd
      FILE src/vadd.cpp
      INCLUDE
        include

>       TARGET kv260

Author: ```suggestion
```

While consering https://discourse.ros.org/t/rep-2008-rfc-ros-2-hardware-acceleration-architecture-and-conventions/22026/8?u=vmayoral, I realized this `TARGET` argument is actually not needed for production, and a  simple leftover from the development process while debugging across different targets.

The embedded (or non-embedded) target itself can be identified by looking at the build arguments provided to `colcon` and/or matched with the firmware deployed and selected in the ROS 2 workspace. In a nutshell, it's not needed.
